llama-index-packs/llama-index-packs-corrective-rag/llama_index/packs/corrective_rag/base.py

```diff
@@ -44,89 +55,161 @@
 )
 
 
-class CorrectiveRAGPack(BaseLlamaPack):
-    def __init__(self, documents: List[Document], tavily_ai_apikey: str) -> None:
-        """Init params."""
+class RetrieveEvent(Event):
+    """Retrieve event (gets retrieved nodes)."""
+
+    retrieved_nodes: List[NodeWithScore]
+
+
+class RelevanceEvalEvent(Event):
+    """Relevance evaluation event (gets results of relevance evaluation)."""
+
+    relevant_results: List[str]
+
+
+class TextExtractEvent(Event):
+    """Text extract event. Extracts relevant text and concatenates."""
+
+    relevant_text: str
+
+
+class QueryEvent(Event):
+    """Query event. Queries given relevant text and search text."""
+
+    relevant_text: str
+    search_text: str
+
+
+class CorrectiveRAGWorkflow(Workflow):
+    @step(pass_context=True)
+    async def ingest(self, ctx: Context, ev: StartEvent) -> Optional[StopEvent]:
+        """Ingest step (for ingesting docs and initializing index)."""
+        documents: Optional[List[Document]] = ev.get("documents")
+        tavily_ai_apikey: Optional[str] = ev.get("tavily_ai_apikey")
+
+        if any(i is None for i in [documents, tavily_ai_apikey]):
+            return None
+
         llm = OpenAI(model="gpt-4")
-        self.relevancy_pipeline = QueryPipeline(
+        ctx.data["relevancy_pipeline"] = QueryPipeline(
```

Comment:
this...is probably ok for now until we have hierarchical workflows? @logan-markewich thoughts?

Reply:
I mean, we could reimplement the pipeline to be a workflow now too

Reply:
and update with the slightly nicer api when it lands

Reply:
@logan-markewich unrelated but one of the original motivations for QPs were the syntatic sugar for common chains e.g. (prompt -> LLM -> output parsing) 

should we do that for workflows?